runtime/pkg/rilltime/rilltime.go

```diff
@@ -0,0 +1,394 @@
+		"PP": "",
+		"PD": "-1d,0d",
+		"PW": "-1W,0W",
+		"PM": "-1M,0M",
+		"PQ": "-1Q,0Q",
+		"PY": "-1Y,0Y",
+	}
+	rillTimeParser = participle.MustBuild[Expression](
+		participle.Lexer(rillTimeLexer),
+		participle.Elide("Whitespace"),
+	)
+	grainMap = map[string]timeutil.TimeGrain{
+		"s": timeutil.TimeGrainSecond,
+		"m": timeutil.TimeGrainMinute,
+		"h": timeutil.TimeGrainHour,
+		"d": timeutil.TimeGrainDay,
+		"D": timeutil.TimeGrainDay,
+		"W": timeutil.TimeGrainWeek,
+		"Q": timeutil.TimeGrainQuarter,
+		"M": timeutil.TimeGrainMonth,
+		"Y": timeutil.TimeGrainYear,
+	}
+)
+
+type Expression struct {
+	Start         *TimeAnchor  `parser:"  @@"`
+	End           *TimeAnchor  `parser:"(',' @@)?"`
+	Modifiers     *Modifiers   `parser:"(':' @@)?"`
+	AtModifiers   *AtModifiers `parser:"('@' @@)?"`
+	isNewFormat   bool
+	grain         *Grain
+	truncateGrain timeutil.TimeGrain
+	isComplete    bool
+	timeZone      *time.Location
+}
+
+type TimeAnchor struct {
+	Grain       *Grain  `parser:"( @@"`
+	AbsDate     *string `parser:"| @AbsoluteDate"`
+	AbsTime     *string `parser:"| @AbsoluteTime"`
+	Earliest    bool    `parser:"| @Earliest"`
+	Now         bool    `parser:"| @Now"`
+	Latest      bool    `parser:"| @Latest"`
+	Watermark   bool    `parser:"| @Watermark)"`
+	Offset      *Grain  `parser:"@@?"`
+	Trunc       *string `parser:"  ('/' @Grain)?"`
+	isoDuration *duration.StandardDuration
+}
+
+type Modifiers struct {
+	Grain         *Grain `parser:"( @@"`
+	CompleteGrain *Grain `parser:"| '|' @@ '|')?"`
+}
+
+type Grain struct {
+	Num   *int   `parser:"@Number?"`
+	Grain string `parser:"@Grain"`
+}
+
+type AtModifiers struct {
+	Offset   *TimeAnchor `parser:"@@?"`
+	TimeZone *string     `parser:"@TimeZone?"`
+}
+
+type EvalOptions struct {
+	Now        time.Time
+	MinTime    time.Time
+	MaxTime    time.Time
+	Watermark  time.Time
+	FirstDay   int
+	FirstMonth int
+}
+
+func Parse(from string) (*Expression, error) {
+	var rt *Expression
+	var err error
+
+	rt, err = ParseISO(from, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if rt == nil {
+		rt, err = rillTimeParser.ParseString("", from)
+		if err != nil {
+			return nil, err
+		}
+		rt.isNewFormat = true
+	}
+
+	rt.timeZone = time.UTC
+	if rt.Modifiers != nil {
+		if rt.Modifiers.Grain != nil {
+			rt.truncateGrain = grainMap[rt.Modifiers.Grain.Grain]
+			rt.grain = rt.Modifiers.Grain
+			// TODO: non-1 grains
+		} else if rt.Modifiers.CompleteGrain != nil {
+			rt.truncateGrain = grainMap[rt.Modifiers.CompleteGrain.Grain]
+			rt.grain = rt.Modifiers.CompleteGrain
+			// TODO: non-1 grains
+			rt.isComplete = true
+		}
+	}
+	if rt.AtModifiers != nil && rt.AtModifiers.TimeZone != nil {
+		var err error
+		rt.timeZone, err = time.LoadLocation(strings.Trim(*rt.AtModifiers.TimeZone, "{}"))
+		if err != nil {
+			return nil, fmt.Errorf("invalid time zone %q: %w", *rt.AtModifiers.TimeZone, err)
+		}
+	}
+
+	if rt.End == nil {
+		rt.End = &TimeAnchor{
+			Now: true,
```

Comment:
Shouldn't the watermark be the default end if not specified?

Reply:
Lemme check with mike today. The docs mention that it should be now. But it probably make sense to do latest/watermark.

Reply:
Got a confirmation from @nishantmonu51 to use watermark.